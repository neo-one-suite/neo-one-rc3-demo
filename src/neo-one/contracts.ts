/* @hash 65b98f0b8f28d3f1c5b60a219e8b8904 */
// tslint:disable
/* eslint-disable */
import { Client } from '@neo-one/client';

import { TestICOSmartContract, TestICOMigrationSmartContract } from './TestICO/types';
import { TestTokenSmartContract, TestTokenMigrationSmartContract } from './TestToken/types';

import { createTestICOSmartContract } from './TestICO/contract';
import { createTestTokenSmartContract } from './TestToken/contract';

export interface Contracts<TClient extends Client = Client> {
  readonly testIco: TestICOSmartContract<TClient>;
  readonly testToken: TestTokenSmartContract<TClient>;
}
// Refer to the MigrationSmartContract documentation at https://neo-one.io/docs/deployment for more information.
export interface MigrationContracts {
  readonly testIco: TestICOMigrationSmartContract;
  readonly testToken: TestTokenMigrationSmartContract;
}

export const createContracts = <TClient extends Client>(client: TClient): Contracts<TClient> => ({
  testIco: createTestICOSmartContract(client),
  testToken: createTestTokenSmartContract(client),
});
