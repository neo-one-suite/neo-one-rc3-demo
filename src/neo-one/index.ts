/* @hash ffe3757c81d0d5ed6e9dbc60d9c1a82c */
// tslint:disable
/* eslint-disable */
export * from './contracts';
export * from './react';
export * from './client';
export * from './TestICO/contract';
export * from './TestICO/types';
export * from './TestICO/manifest';
export * from './TestToken/contract';
export * from './TestToken/types';
export * from './TestToken/manifest';
