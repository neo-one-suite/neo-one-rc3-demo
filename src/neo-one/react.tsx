/* @hash 4942a0fa75db6e9c9bb6274c9f85a5f4 */
// tslint:disable
/* eslint-disable */
import { Client, DeveloperClients, DeveloperTools } from '@neo-one/client';
import * as React from 'react';
import { Contracts } from './contracts';
import { createClient, createDeveloperClients } from './client';
import { createTestICOSmartContract } from './TestICO/contract';
import { createTestTokenSmartContract } from './TestToken/contract';

export interface WithClients<TClient extends Client> {
  readonly client: TClient;
  readonly developerClients: DeveloperClients;
  readonly host?: string;
}
export type ContractsWithClients<TClient extends Client> = Contracts & WithClients<TClient>;

const Context = React.createContext<ContractsWithClients<any>>(undefined as any);

export type ContractsProviderProps<TClient extends Client> = Partial<WithClients<TClient>> & {
  readonly children?: React.ReactNode;
};
export const ContractsProvider = <TClient extends Client>({
  client: clientIn,
  developerClients: developerClientsIn,
  host,
  children,
}: ContractsProviderProps<TClient>) => {
  const client = clientIn === undefined ? createClient(host) : clientIn;
  const developerClients = developerClientsIn === undefined ? createDeveloperClients(host) : developerClientsIn;
  DeveloperTools.enable({ client, developerClients });

  return (
    <Context.Provider
      value={{
        client,
        developerClients,
        testIco: createTestICOSmartContract(client),
        testToken: createTestTokenSmartContract(client),
      }}
    >
      {children}
    </Context.Provider>
  );
};

export interface WithContractsProps<TClient extends Client> {
  readonly children: (contracts: ContractsWithClients<TClient>) => React.ReactNode;
}
export const WithContracts = <TClient extends Client>({ children }: WithContractsProps<TClient>) => (
  <Context.Consumer>{children}</Context.Consumer>
);

export const useContracts = <TClient extends Client>(): ContractsWithClients<TClient> => React.useContext(Context);
