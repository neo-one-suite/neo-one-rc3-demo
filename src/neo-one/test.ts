/* @hash 14a0b1942e9242f99637c7235baa6c13 */
// tslint:disable
/* eslint-disable */
import { createWithContracts, TestOptions, WithContractsOptions } from '@neo-one/smart-contract-test';
import { Contracts } from './contracts';
import * as path from 'path';

export const withContracts: (
  test: (contracts: Contracts & TestOptions) => Promise<void>,
  options?: WithContractsOptions,
) => Promise<void> = createWithContracts([
  { name: 'TestICO', filePath: path.resolve(__dirname, '../../neo-one/contracts/TestICO.ts') },
  { name: 'TestToken', filePath: path.resolve(__dirname, '../../neo-one/contracts/TestToken.ts') },
]);
