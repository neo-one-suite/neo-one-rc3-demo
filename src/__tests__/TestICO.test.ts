import { common, crypto } from '@neo-one/client-common';
import { Hash160, privateKeyToAddress } from '@neo-one/client';
import BigNumber from 'bignumber.js';
import { withContracts } from '../neo-one/test';

jest.setTimeout(10000);

const MINTER = {
  PRIVATE_KEY:
    '536f1e9f0466f6cd5b2ea5374d00f038786daa0f0e892161d6b0cb4d6b154740',
  PUBLIC_KEY:
    '03463b7a0afc41ff1f6a386190f99bafd1deca48f4026aeac95435731af278cb7d',
};

describe('TestICO', () => {
  test('smart contract', async () => {
    await withContracts(
      async ({
        client,
        networkName,
        testIco: smartContract,
        masterAccountID,
      }) => {
        crypto.addPublicKey(
          common.stringToPrivateKey(MINTER.PRIVATE_KEY),
          common.stringToECPoint(MINTER.PUBLIC_KEY),
        );
        const deployResult = await smartContract.deploy(
          masterAccountID.address,
          new BigNumber(Math.round(Date.now() / 1)),
        );
        const deployReceipt = await deployResult.confirmed({ timeoutMS: 2500 });
        if (deployReceipt.result.state !== 'HALT') {
          throw new Error(deployReceipt.result.message);
        }

        expect(deployReceipt.result.gasConsumed.toString()).toMatchSnapshot(
          'deploy consumed',
        );
        expect(deployReceipt.result.value).toBeTruthy();

        const [
          nameResult,
          decimalsResult,
          symbolResult,
          minter,
        ] = await Promise.all([
          smartContract.name(),
          smartContract.decimals(),
          smartContract.symbol(),
          client.providers.memory.keystore.addUserAccount({
            network: networkName,
            name: 'minter',
            privateKey: MINTER.PRIVATE_KEY,
          }),
        ]);
        expect(nameResult).toEqual('TestToken');
        expect(decimalsResult.toString()).toEqual('8');
        expect(symbolResult).toEqual('TT');

        const [initialTotalSupply, transferResult] = await Promise.all([
          smartContract.totalSupply(),
          client.transfer(
            [
              {
                amount: new BigNumber(10000),
                asset: Hash160.NEO,
                to: privateKeyToAddress(MINTER.PRIVATE_KEY),
              },
              {
                amount: new BigNumber(154460781), // TODO: why is the GAS required so large?
                asset: Hash160.GAS,
                to: privateKeyToAddress(MINTER.PRIVATE_KEY),
              },
            ],
            {
              from: masterAccountID,
            },
          ),
        ]);
        expect(initialTotalSupply.toString()).toEqual('0');

        await transferResult.confirmed({ timeoutMS: 2500 });

        const firstMint = new BigNumber('10');
        const mintResult = await smartContract.mintTokens({
          from: minter.userAccount.id,
          sendTo: [
            {
              amount: firstMint,
              asset: common.nativeScriptHashes.NEO,
            },
          ],
        });

        const mintReceipt = await mintResult.confirmed({ timeoutMS: 2500 });
        expect(mintReceipt.result.gasConsumed.toString()).toMatchSnapshot(
          'mint consumed',
        );
        expect(mintReceipt.result.value).toBeUndefined();
        expect(mintReceipt.events).toHaveLength(3);
        const event = mintReceipt.events[2];
        expect(event.name).toEqual('Transfer');
        expect(event.parameters.from).toBeUndefined();
        expect(event.parameters.to).toEqual(minter.userAccount.id.address);
        if (event.parameters.amount === undefined) {
          expect(event.parameters.amount).toBeTruthy();
          throw new Error('For TS');
        }
        const firstBalance = firstMint.times(10).toString();
        expect(
          event.parameters.amount.times(new BigNumber(100000000)).toString(),
        ).toEqual(firstBalance);

        const [minterBalance, mintTotalSupply] = await Promise.all([
          smartContract.balanceOf(minter.userAccount.id.address),
          smartContract.totalSupply(),
        ]);

        expect(
          minterBalance.times(new BigNumber(100000000)).toString(10),
        ).toEqual(firstBalance);
        expect(
          mintTotalSupply.times(new BigNumber(100000000)).toString(10),
        ).toEqual(firstBalance);
      },
      { deploy: false },
    );
  });
});
