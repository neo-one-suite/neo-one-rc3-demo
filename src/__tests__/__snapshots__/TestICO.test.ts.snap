// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestICO smart contract: deploy consumed 1`] = `"47760930"`;

exports[`TestICO smart contract: mint consumed 1`] = `"132427990"`;
